File: identity-service/src/main/java/com/solarease/service/AuthService.java
package com.solarease.service;

import com.solarease.dto.AuthResponse;
import com.solarease.dto.LoginRequest;
import com.solarease.dto.RegisterRequest;
import com.solarease.dto.VerifyOtpRequest;
import com.solarease.entity.User;
import com.solarease.exception.BadRequestException;
import com.solarease.exception.ResourceNotFoundException;
import com.solarease.repository.UserRepository;
import com.solarease.security.JwtTokenProvider;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Slf4j
@Transactional
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider jwtTokenProvider;
    private final OtpService otpService;
    private final EmailService emailService;

    public AuthService(UserRepository userRepository, 
                      PasswordEncoder passwordEncoder,
                      JwtTokenProvider jwtTokenProvider,
                      OtpService otpService,
                      EmailService emailService) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.jwtTokenProvider = jwtTokenProvider;
        this.otpService = otpService;
        this.emailService = emailService;
    }

    public AuthResponse login(LoginRequest request) {
        log.info("Login attempt for email: {}", request.getEmail());
        
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new ResourceNotFoundException("User not found with email: " + request.getEmail()));

        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new BadRequestException("Invalid email or password");
        }

        if (!user.getIsActive()) {
            throw new BadRequestException("User account is inactive");
        }

        if (!user.getIsEmailVerified()) {
            throw new BadRequestException("Email not verified. Please verify your email first.");
        }

        String accessToken = jwtTokenProvider.generateAccessToken(user.getUuid(), user.getEmail());
        String refreshToken = jwtTokenProvider.generateRefreshToken(user.getUuid(), user.getEmail());

        AuthResponse.UserDto userDto = buildUserDto(user);

        log.info("User {} logged in successfully", user.getEmail());
        
        return AuthResponse.builder()
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .expiresIn(jwtTokenProvider.getExpirationTime())
                .user(userDto)
                .build();
    }

    public AuthResponse register(RegisterRequest request) {
        log.info("Register attempt for email: {}", request.getEmail());
        
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BadRequestException("Email already registered");
        }

        if (userRepository.existsByUsername(request.getUsername())) {
            throw new BadRequestException("Username already taken");
        }

        User newUser = User.builder()
                .uuid(java.util.UUID.randomUUID().toString())
                .email(request.getEmail())
                .username(request.getUsername())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .firstName(request.getFirstName())
                .lastName(request.getLastName())
                .phone(request.getPhone())
                .role(User.UserRole.CLIENT)
                .isActive(true)
                .isEmailVerified(false)
                .build();

        User savedUser = userRepository.save(newUser);

        // Generate and send OTP
        otpService.generateAndSendOtp(savedUser);

        log.info("User {} registered. OTP sent to email", savedUser.getEmail());
        
        return AuthResponse.builder()
                .message("User registered successfully. Please verify your email with the OTP code sent.")
                .build();
    }

    public AuthResponse verifyOtp(VerifyOtpRequest request) {
        log.info("OTP verification attempt for email: {}", request.getEmail());

        // Validate OTP
        otpService.validateOtp(request.getEmail(), request.getOtpCode());

        // Mark email as verified
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));

        user.setIsEmailVerified(true);
        userRepository.save(user);

        // Send welcome email
        emailService.sendWelcomeEmail(user.getEmail(), user.getFirstName());

        log.info("Email verified for user: {}", request.getEmail());

        return AuthResponse.builder()
                .message("Email verified successfully! You can now login.")
                .email(request.getEmail())
                .build();
    }

    private AuthResponse.UserDto buildUserDto(User user) {
        return AuthResponse.UserDto.builder()
                .id(user.getId())
                .uuid(user.getUuid())
                .email(user.getEmail())
                .username(user.getUsername())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .role(user.getRole().toString())
                .build();
    }
}

File: SolarEase Web App Design (2)/src/app/services/authService.ts
import api from "./api";

export interface LoginRequest {
  email: string;
  password: string;
}

export interface AuthUser {
  id: number;
  uuid: string;
  email: string;
  username: string;
  firstName: string;
  lastName: string;
  phone: string;
  role: string;
  isEmailVerified: boolean;
  isActive: boolean;
  createdAt: string;
}

export interface InstallerOption {
  id: number;
  uuid: string;
  username: string;
  email: string;
  firstName: string;
  lastName: string;
  phone: string;
  role: string;
  isEmailVerified: boolean;
  isActive: boolean;
  createdAt: string;
}

export interface AuthResponse {
  accessToken: string;
  refreshToken: string;
  expiresIn: number;
  user: AuthUser;
  message?: string;
  email?: string;
  otpCode?: string;
}

export interface UpdateProfileRequest {
  firstName: string;
  lastName: string;
  phone: string;
}

export interface ChangePasswordRequest {
  currentPassword: string;
  newPassword: string;
  confirmPassword: string;
}

export interface RegisterRequest {
  username?: string;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  password: string;
  userRole?: "CLIENT" | "INSTALLER";
}

export interface VerifyOtpRequest {
  email: string;
  otpCode: string;
  invitationToken?: string;
}

export interface ResendOtpRequest {
  email: string;
}

export interface CreateInstallerRequest {
  firstName: string;
  lastName: string;
  email: string;
  phone?: string;
  password: string;
}

export interface UpdateInstallerRequest {
  firstName?: string;
  lastName?: string;
  email?: string;
  phone?: string;
  password?: string;
  isActive?: boolean;
}

const authService = {
  login: (data: LoginRequest) =>
    api.post<AuthResponse>("/auth/login", data).then((r) => r.data),

  register: (data: RegisterRequest) =>
    api.post<AuthResponse>("/auth/register", data).then((r) => r.data),

  verifyOtp: (data: VerifyOtpRequest) =>
    api.post<AuthResponse>("/auth/verify-otp", data).then((r) => r.data),

  resendOtp: (data: ResendOtpRequest) =>
    api.post<AuthResponse>("/auth/resend-otp", data).then((r) => r.data),

  getProfile: () =>
    api.get<AuthResponse>("/auth/me").then((r) => r.data),

  getInstallers: () =>
    api.get<InstallerOption[]>("/auth/installers").then((r) => r.data),

  createInstaller: (data: CreateInstallerRequest) =>
    api.post<InstallerOption>("/auth/installers", data).then((r) => r.data),

  updateInstaller: (installerUuid: string, data: UpdateInstallerRequest) =>
    api.put<InstallerOption>(`/auth/installers/${installerUuid}`, data).then((r) => r.data),

  deleteInstaller: (installerUuid: string) =>
    api.delete<void>(`/auth/installers/${installerUuid}`).then((r) => r.data),

  updateProfile: (data: UpdateProfileRequest) =>
    api.put<AuthResponse>("/auth/me", data).then((r) => r.data),

  changePassword: (data: ChangePasswordRequest) =>
    api.put<{ message: string }>("/auth/me/password", data).then((r) => r.data),

  refreshToken: (refreshToken: string) =>
    api.post<AuthResponse>("/auth/refresh", { refreshToken }).then((r) => r.data),

  logout: () => {
    localStorage.removeItem("accessToken");
    localStorage.removeItem("refreshToken");
    localStorage.removeItem("user");
  },
};

export default authService;
